File: SafetyProgram.UI.Views/MainView/Default/MainViewModel.cs
using System;
using System.ComponentModel;
using System.Windows.Controls;
using Fluent;
using SafetyProgram.Base;
using SafetyProgram.Core;
using SafetyProgram.Core.Models;

namespace SafetyProgram.UI.Views.MainView.Default
{
    /// <summary>
    /// Defines a standard implementation of an <code>IMainViewModel</code>
    /// </summary>
    public sealed class MainViewModel : 
        IMainViewModel
    {
        public MainViewModel(IApplicationKernel model,
            Ribbon ribbonView,
            Func<IDocument, Control> documentViewFactory)
        {
            Helpers.NullCheck(model,
                ribbonView,
                documentViewFactory);

            this.RibbonView = ribbonView;

            model.DocumentChanged +=
                (s, e) =>
                {
                    this.ContentView = e.NewProperty == null ?
                        null :
                        documentViewFactory(e.NewProperty);
                };
            this.ContentView = model.Document == null ? 
                null : 
                documentViewFactory(model.Document);
        }

        /// <summary>
        /// Get the ribbon view.
        /// </summary>
        public Ribbon RibbonView { get; private set; }

        private Control contentView;

        /// <summary>
        /// Get the content view.
        /// </summary>
        public Control ContentView 
        {
            get { return contentView; }
            set
            {
                contentView = value;
                PropertyChanged.Raise(this, "ContentView");
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;
    }
}

File: TEMP.ViewModels/MainViewModel.fs
namespace TEMP.ViewModels

open System
open System.ComponentModel
open SafetyProgram.UI.Views.MainView.Default
open SafetyProgram.Core
open Fluent

// Defines a standard implementation of a <code>IMainViewModel</code>
type FSViewModel(model : IApplicationKernel, ribbon, documentViewFactory) as this = 
    // Event that's called when property changes.
    let propertyChangedEvent = new Event<_,_>()

    let mutable contentView = documentViewFactory(model.Document)

    // Add hook that updates viewmodel when underlying model changes
    do
        model.DocumentChanged.Add(fun e -> 
                contentView<-documentViewFactory(e.NewProperty)
                propertyChangedEvent.Trigger(
                    this, 
                    new PropertyChangedEventArgs("ContentView")))  
    
    // Viewmodel implementation
    interface IMainViewModel with
        member this.RibbonView = ribbon        
        member this.ContentView = contentView
        [<CLIEvent>]
        member this.PropertyChanged = propertyChangedEvent.Publish